nodes/generators_extended/torus2D.py

```diff
@@ -0,0 +1,273 @@
+        self.inputs.new('StringsSocket', "n2").prop_name = 'torus_n2'
+        self.inputs.new('StringsSocket', "rP").prop_name = 'torus_rP'
+
+        self.outputs.new('VerticesSocket', "Vertices")
+        self.outputs.new('StringsSocket',  "Edges")
+        self.outputs.new('StringsSocket',  "Polygons")
+
+    def draw_buttons(self, context, layout):
+        layout.prop(self, "Separate", text="Separate")
+        layout.prop(self, 'mode', expand=True)
+
+    def process(self):
+        # return if no outputs are connected
+        if not any(s.is_linked for s in self.outputs):
+            return
+
+        # input values lists (single or multi value)
+        # list of MAJOR or EXTERIOR radii
+        input_RR = self.inputs["R"].sv_get()[0]
+        # list of MINOR or INTERIOR radii
+        input_rr = self.inputs["r"].sv_get()[0]
+        # list of number of MAJOR sections : RADIAL
+        input_n1 = self.inputs["n1"].sv_get()[0]
+        # list of number of MINOR sections : CIRCULAR
+        input_n2 = self.inputs["n2"].sv_get()[0]
+        # list of RADIAL phases
+        input_rp = self.inputs["rP"].sv_get()[0]
+
+        # sanitize the input values
+        input_RR = list(map(lambda x: max(0, x), input_RR))
+        input_rr = list(map(lambda x: max(0, x), input_rr))
+        input_n1 = list(map(lambda x: max(3, int(x)), input_n1))
+        input_n2 = list(map(lambda x: max(2, int(x)), input_n2))
+
+        # convert input radii values to MAJOR/MINOR, based on selected mode
+        if self.mode == 'EXT_INT':
+            # convert radii from EXTERIOR/INTERIOR to MAJOR/MINOR
+            # (extend radii lists to a matching length before conversion)
+            input_RR, input_rr = match_long_repeat([input_RR, input_rr])
+            input_R = list(map(lambda x, y: (x+y)*0.5, input_RR, input_rr))
+            input_r = list(map(lambda x, y: (x-y)*0.5, input_RR, input_rr))
+        else:  # values already given as MAJOR/MINOR radii
+            input_R = input_RR
+            input_r = input_rr
+
+        parameters = match_long_repeat([input_R, input_r, input_n1, input_n2, input_rp])
+
+        if self.outputs['Vertices'].is_linked:
+            vertList = []
+            for R, r, n1, n2, p in zip(*parameters):
+                verts = torus_verts(R, r, n1, n2, p, self.Separate)
+                vertList.append(verts)
+            self.outputs['Vertices'].sv_set(vertList)
+
+        if self.outputs['Edges'].is_linked:
+            edgeList = []
+            for R, r, n1, n2, _ in zip(*parameters):
+                edges = torus_edges(n1, n2)
+                edgeList.append(edges)
+            self.outputs['Edges'].sv_set(edgeList)
+
+        if self.outputs['Polygons'].is_linked:
+            polyList = []
+            for R, r, n1, n2, _ in zip(*parameters):
+                polys = torus_polygons(n1, n2)
+                polyList.append(polys)
+            self.outputs['Polygons'].sv_set(polyList)
+
```

Comment:
```python
        E, P = self.outputs['Edges'], self.outputs['Polygons']
        for socket, func in [(E, torus_edges),(P, torus_polygons)]:
            if socket.is_linked:
                socket.sv_set([func(n1, n2) for _, _, n1, n2, _ in zip(*parameters)])
```
Maybe that's taking it too far, but that's how I would write these..because they are identical in signature.